pkg/binding/bindingpolicy-resolver.go

```diff
@@ -95,6 +98,7 @@ type BindingPolicyResolver interface {
 	// The given destinations set is expected not to be mutated during and
 	// after this call by the caller.
 	// If no resolution is associated with the given key, an error is returned.
+	// Must not be called concurrently with any method that adds resolutions.
```

Comment:
A. Why do we need this constraint
B. Can it really be enforced? What if two different BindingPolicies are processed concurrently?

Reply:
Sorry, I had not written the comment precisely enough. I have now revised it, to clarify that the concern is only with concurrent activity wrt the same BindingPolicy name.

The value it brings is making the implementation correct. Without that constraint or something else about usage, the implementation has a concurrency bug: the implementation fetches a resolution out of the resolver's map and then releases the lock on that map before making any change in the resolution; this opens up the possibility that concurrent activity migh remove that resolution and introduce a replacement (a resolution under the same key).

Reply:
I originally thought that the concurrency problem is not conforming to the constraint of not directly accessing fields within the returned resolutions.

If a mutating call such as `SetDestinations` gets to fetching the resolution, then before executing the mutating method a reconciliation of the same key swapped the resolution in memory, then the mutating call would be irrelevant to the new resolution. It would be executed, it may alert subscribers of change, but once the change is reconciled, the state remains correct (redundant consumption at worst).

A simple solution to this problem may be adding a "valid" field and performing the double-checked locking pattern.

Reply:
I am looking at multiple concurrency issues.

I am not sure that I would use the word "enforce". The current code _obeys_ this constraint.

Reply:
How does it obey it? Also, do you agree with what I said in the 2nd paragraph?

Reply:
Both of your questions have the same answer. There never are two concurrent invocations of `BindingPolicyResolver.SetDestinations` for the same BindingPolicy name because it is only called from `Controller.syncBindingPolicy` and, because of the way the controller pattern works, there are never two concurrent calls on that with the same BindingPolicy name.

Reply:
I don't mean two concurrent calls to the same method specifically. What I meant is the only concurrency bug in the code is instances of accessing a field directly within a returned resolution. Instead, all should be accessed by getters and setters.

The issue of working with outdated binding-resolutions in case one is fetched and then replaced in memory when the lock is released, is not a bug, but a choice. Its price is redundant actions on the orphaned resolution and possibly a redundant sync check. I took this option because I thought it is statistically worth it.

Reply:
In this PR, I am addressing more than the concurrency bugs called out in issue #2418 . Looking at the current definition of BindingPolicyResolver (https://github.com/kubestellar/kubestellar/blob/f30a85da33d6b9df8f9629ff27ffce74326b4e56/pkg/binding/bindingpolicy-resolver.go#L33-L130), I noticed that this interface does not forbid concurrent updates to the same BindingPolicy. In particular, it does not forbid concurrent updates that conflict. Sure, fixing the bugs called out in #2418 will prevent the code from panicing or failing a mechanical data race checker, but there is still the semantic question of what happens if two concurrent goroutines make conflicting calls on a BindingPolicyResolver. So I tried adding comments explaining why the code (all of the code in the repo taken together) is correct. I am not happy with what is in this PR, because some of the comments are inside method implementations yet take into account what the only client of this abstraction does with it --- thus breaking the abstraction. I hope to do better in follow-on PRs.